goaviatrix/edge_megaport.go

```diff
@@ -0,0 +1,262 @@
+	Longitude                          float64             `json:"longitude"`
+	WanPublicIp                        string              `json:"public_ip"`
+	PrivateIP                          string              `json:"private_ip"`
+	RxQueueSize                        string              `json:"rx_queue_size"`
+	State                              string              `json:"vpc_state"`
+	InterfaceList                      []MegaportInterface `json:"interfaces"`
+	DnsProfileName                     string              `json:"dns_profile_name"`
+	EnableNat                          string              `json:"enable_nat"`
+	SnatMode                           string              `json:"snat_target"`
+	EnableAutoAdvertiseLanCidrs        bool                `json:"auto_advertise_lan_cidrs"`
+	InterfaceMapping                   []InterfaceMapping  `json:"interface_mapping"`
+}
+
+type EdgeMegaportListResp struct {
+	Return  bool               `json:"return"`
+	Results []EdgeMegaportResp `json:"results"`
+	Reason  string             `json:"reason"`
+}
+
+type MegaportInterface struct {
+	Index         int     `json:"index"`
+	Type          string  `json:"type"`
+	Name          string  `json:"ifname,omitempty"`
+	PublicIp      string  `json:"public_ip,omitempty"`
+	Tag           string  `json:"tag,omitempty"`
+	Dhcp          bool    `json:"dhcp,omitempty"`
+	IpAddr        string  `json:"ipaddr,omitempty"`
+	GatewayIp     string  `json:"gateway_ip,omitempty"`
+	DnsPrimary    string  `json:"dns_primary,omitempty"`
+	DnsSecondary  string  `json:"dns_secondary,omitempty"`
+	SubInterfaces []*Vlan `json:"subinterfaces,omitempty"`
+	VrrpState     bool    `json:"vrrp_state,omitempty"`
+	VirtualIp     string  `json:"virtual_ip,omitempty"`
+}
+
+type CreateEdgeMegaportResp struct {
+	Return bool   `json:"return"`
+	Result string `json:"results"`
+	Reason string `json:"reason"`
+}
+
+func (c *Client) CreateEdgeMegaport(ctx context.Context, edgeMegaport *EdgeMegaport) error {
+	edgeMegaport.Action = "create_megaport_instance"
+	edgeMegaport.CID = c.CID
+	edgeMegaport.NoProgressBar = true
+
+	interfaces, err := json.Marshal(edgeMegaport.InterfaceList)
+	if err != nil {
+		return err
+	}
+
+	edgeMegaport.Interfaces = b64.StdEncoding.EncodeToString(interfaces)
+
+	if len(edgeMegaport.VlanList) == 0 {
+		edgeMegaport.VlanList = []*EdgeMegaportVlan{}
+	}
+
+	vlan, err := json.Marshal(edgeMegaport.VlanList)
+	if err != nil {
+		return err
+	}
+
+	edgeMegaport.Vlan = b64.StdEncoding.EncodeToString(vlan)
+
+	var data CreateEdgeMegaportResp
+
+	err = c.PostAPIContext2(ctx, &data, edgeMegaport.Action, edgeMegaport, BasicCheck)
+	if err != nil {
+		return err
+	}
+
+	fileName := edgeMegaport.ZtpFileDownloadPath + "/" + edgeMegaport.GwName + "-" + edgeMegaport.SiteId + "-cloud-init.txt"
+
+	outFile, err := os.Create(fileName)
+	if err != nil {
+		return err
+	}
+
+	_, err = outFile.WriteString(data.Result)
+	if err != nil {
+		return err
+	}
```

Comment:
Is the user supposed to do sth with this file?
What happens if terraform runs in a CD pipeline?
Would it make sense to expose such things as computed properties?

Reply:
This is the cloud_init file which is required to bring up the Megaport Edge Gateway. The user's will have to onboard a megaport account, create MVE with internet connection enabled on the Megaport portal. Cloud_init file is required to successfully order an MVE which will be provided by the TF resource.

Reply:
I see that the same logic is used in edge_equinix, so it seems to be accepted.
I still believe that it would be better to expose the cloud-init contents as a computed property.
Then the user can still write it to a local file, or grab it from their CICD pipeline.